### ISSUE TITLE
bug: Content Layer v5 content collections do not automatically exclude underscore pages

### ISSUE THREAD

**Issue by techfg** [vernak2539/astro-rehype-relative-markdown-links#75]
In #25, the plugin was updated to automatically exclude resolving markdown references for pages that contain an underscore in their name and/or parent names.  According to the [v4 docs](https://github.com/withastro/docs/issues/10300), this aligned with the way Astro v4 worked.  However, in v5, the [Content Layer](https://github.com/withastro/astro/issues/12730) no longer excludes content collection pages that contain an underscore.

To address this change, there are a few options:

1. Continue to exclude underscore - This would be consistent for v4, but not for v5
2. Do not exclude underscore pages by eliminating the current check - This would be consistent with v5 but be breaking change for v4
3. Add an `ignore` configuration property that is a RegEx to allow excluding any pattern

Currently, while underscores are skipped in the plugin, the final html still contains a link that is invalid.  For example, `[My Stuff](./_mystuff.md]` would result in `/docs/_mystuff.md` which would be an invalid link in v4 & v5 but in v4, while the `/docs/_mystuff` would not exist by default, in v5 it would by default.  It all comes down to the pattern that is set for the [glob loader](https://docs.astro.build/en/guides/content-collections/#built-in-loaders) in v5.

Given above, I think option 2 or 3 makes the most sense with a strong lean towards option 2 since currently, a link will exist in the final html anyway even if the plugin skips it.  I do think there is something to be said for adding an `ignore` configuration property but again, there's going to be a link anyway so why not just transform it if it does exist.  If the goal is to not have invalid links in the final html, there are tools that validate links that can be used to catch those.

@vernak2539 - Thoughts?

**Comment by techfg** [comment]
Thanks for putting together the initial PR @vernak2539!

Here's my thoughts:

1. I'd be hesitant to add astro version options in general as over-time, it just makes the code increasingly different to maintain/support
2. From what I see in the Astro ecosystem, a lot of plugins/integrations only support current Astro version or at most last two versions
3. This is the first issue that has arisen from v2 to v5 that requires this plugin to potentially behave different based on Astro version
4. Even in v4, the code that currently skips transforming underscore (`_`) links results in a link being emitted to the final HTML, it just points to `/blog/_mypage.mdx` instead of `/blog/mypage` but neither link would actually work in v4, whereas in v5, the second link would unless the user configured the glob pattern to ignore underscore links.  Given this, since there's a link anyway, ignoring the transformation doesn't really do much good anyway in v4 and in v5, actually creates a potential issue.

All that said and with other content layer changes that are going to be necessary (see #74), my thoughts are the following:

1. Create & document an official position on supported versions - my recommendation would be current version only or last two major only.  If current version only, we would release a v1 of the plugin that would be the last for v4 and then release a 2.0 for v5.  This is a common pattern in libraries/plugins to ensure they only support current currents of dependent products while still providing an option for prior versions for those that need them.  
2. If supporting 2 most recent, either drop the current underscore check completely or add an `ignore` option.  The ignore option could default to `_` for backwards compat but be adjustable to support those that are taking advantage of `glob` patterns in their collections.  Since a link is going to be generated anyway and it may or may not be valid, I don't think transforming it in all cases and dropping the `shouldProcessFile` concept would be a bad thing.

In short:
1. Official support for current Astro major only
2. Release a v1 for v4 and release a v2 for v5 - We could still fix/update v1 for v4 if it became necessary but it would receive features
3. In the v2 codebase, remove the `shouldProcessFile` check entirely.  If someone doesn't want a particular link to get transformed, they should remove from their MDX because there's going to be a link there regardless, just a matter of whether or not its a dead link.

Thoughts?

**Comment by techfg** [comment]
Yeah, I hear ya about not being ready for a v1, there are still some edge cases to be worked out.

There really is next to zero value in "skipping" transformation since a url will get emitted regardless of whether or not we transform and it may or may not be a valid url so I do think there's something to be said for `Option 2` from the OP.

Short of that, I think adding an `ignore` option, defaulting to v5 behavior meaning `ignore: []` as the default, would provide a solution for anyone that is relying on the plugin current behavior of skipping `_` and also allow for the v5 glob pattern semantics.  If we go this route, we might want to have an `ignore` at both the `global` config level and within each collection so make config easier (we could enforce both).  My thinking is that for a site that has a lot of collections with complex ignores that different across collections, what may be ignored in one collection might not be in another.  Maybe I'm overthinking but seems to make sense :)
